src/test/shared/clients/defaultEcsClient.test.ts

```diff
@@ -0,0 +1,237 @@
+            const targetArr2 = ['arn4', 'arn5', 'arn6']
+            testClient.listTaskDefinitionsResponses = [
+                {
+                    taskDefinitionArns: targetArr1,
+                    nextToken: 'there i go, turn the page'
+                },
+                {
+                    taskDefinitionArns: targetArr2
+                }
+            ]
+            const iterator = testClient.listTaskDefinitions()
+            const arr = []
+            for await (const item of iterator) {
+                arr.push(item)
+            }
+            assert.deepStrictEqual(targetArr1.concat(targetArr2), arr)
+        })
+
+        it('handles errors', async () => {
+            testClient.listTaskDefinitionsResponses = new Error() as AWSError
+            try {
+                const iterator = testClient.listTaskDefinitions()
+                const arr = []
+                for await (const item of iterator) {
+                    arr.push(item)
+                }
+                assert.ok(false)
+            } catch (err) {
+                assert.ok(true)
+            }
+        })
+    })
+})
+
+class TestEcsClient extends DefaultEcsClient {
+
+    public listClustersResponses: ECS.ListClustersResponse[] | AWSError = [{}]
+
+    public listServicesResponses: ECS.ListServicesResponse[] | AWSError = [{}]
+
+    public listTaskDefinitionsResponses: ECS.ListTaskDefinitionsResponse[] | AWSError = [{}]
+
+    private pageNum: number = 0
+
+    public constructor(
+        regionCode: string = 'us-weast-1'
+    ) {
+        super(regionCode)
+    }
+
+    protected async invokeListClusters(request: ECS.ListClustersRequest)
+        : Promise<ECS.ListClustersResponse> {
+        const responseDatum
+            = this.getResponseDatum<ECS.ListClustersResponse>(this.listClustersResponses, request.nextToken)
+
+        if (responseDatum instanceof Error) {
+            throw responseDatum
+        } else {
+            return responseDatum
+        }
+    }
+
+    protected async invokeListServices(request: ECS.ListServicesRequest)
+        : Promise<ECS.ListServicesResponse> {
+        const responseDatum
+            = this.getResponseDatum<ECS.ListServicesResponse>(this.listServicesResponses, request.nextToken)
+
+        if (responseDatum instanceof Error) {
+            throw responseDatum
+        } else {
+            return responseDatum
+        }
+    }
+
+    protected async invokeListTaskDefinitions(request: ECS.ListTaskDefinitionsRequest)
+        : Promise<ECS.ListTaskDefinitionsResponse> {
+        const responseDatum =
+            this.getResponseDatum<ECS.ListTaskDefinitionsResponse>(this.listTaskDefinitionsResponses, request.nextToken)
+
+        if (responseDatum instanceof Error) {
+            throw responseDatum
+        } else {
+            return responseDatum
+        }
+    }
+
+    protected async createSdkClient(): Promise<ECS> {
+        return new ECS()
```

Comment:
To ensure we don't accidentally use the actual client in our tests, you can get away with 
`return {} as any as ECS`